feat: add ability to react on messages(SQSERVICES-1734)

arjita-mitra opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1734" title="SQSERVICES-1734" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />SQSERVICES-1734</a>  [Web] Reactions on Messages
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - add reaction on messages

- The **PR Description**
  - As a Wire user, I would like to have a wide range of reactions (emojis) at my disposal. So that I can use them to express myself in messages and reactions to messages.
---

### Testing

#### How to Test

- added unit testing

##### References
1. https://wearezeta.atlassian.net/browse/SQSERVICES-1734
2. https://www.npmjs.com/package/emoji-picker-react

arjita-mitra committed:
feat: add ability to react on messages

arjita-mitra committed:
feat: create util for transformReactionObj
replace thums up and heart emoji with image
image css
create list reactions component

thisisamir98 committed:
feat: Add emoji title utils (#14827)

arjita-mitra committed:
feat: convert emoji to unicode

arjita-mitra committed:
feat: integrate react functioanlity with reactions button
make reactions button
reactions buttons css
create reaction emoji image component

arjita-mitra committed:
test: unit test message reactions

arjita-mitra committed:
refactor: move message reactions related components under message reactions directory

arjita-mitra committed:
refactor: improve transformReactionObj function

call transformReactionObj from reactions list

arjita-mitra committed:
refactor: update test cases

arjita-mitra committed:
refactor: remove test case

arjita-mitra committed:
refactor: remove unused code
fix: review comment

arjita-mitra committed:
Merge branch 'dev' of https://github.com/wireapp/wire-webapp into feature/message-reaction

arjita-mitra committed:
chore: update yarn lock

arjita-mitra committed:
Merge branch 'feature/SQSERVICES-1864' of https://github.com/wireapp/wire-webapp into feature/message-reaction

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx:
```diff
@@ -62,14 +64,15 @@ export interface ContentMessageProps extends Omit<MessageActions, 'onClickResetS
   quotedMessage?: ContentMessage;
   selfId: QualifiedId;
   isMsgElementsFocusable: boolean;
+  messageRepository: MessageRepository;
```
Humm the content message component should not know about the `messageRepository`. It's doesn't belong in this layer. 
The `MessageWrapper` on the other hand already knows about the `MessageRepository` and should be the one calling the method. 

Instead this component should be given a  callback `onClickReaction` and the parent (`MessageWrapper`) should make the call :)

atomrc commented on src/script/conversation/MessageRepository.ts:
```diff
@@ -786,6 +786,12 @@ export class MessageRepository {
+      window.setTimeout(() => this.sendReaction(conversationEntity, message_et, reaction), 100);
+    }
+  }
+
```
Why not making the sendReaction a public method? Why the timeout?

arjita-mitra replied:
@atomrc I am not sure about that, its the same function that we used for toggleLike I used here. I had to change the reaction type to be string. since it was an existing working code so i didn't change it

arjita-mitra replied:
/**
   * Toggle like status of message.
   *
   * @param conversationEntity Conversation entity
   * @param message_et Message to react to
   */
  public toggleLike(conversationEntity: Conversation, message_et: ContentMessage): void {
    if (!conversationEntity.removed_from_conversation()) {
      const reaction = message_et.is_liked() ? ReactionType.NONE : ReactionType.LIKE;
      message_et.is_liked(!message_et.is_liked());

      window.setTimeout(() => this.sendReaction(conversationEntity, message_et, reaction), 100);
    }
  }

atomrc replied:
I'd say the timeout def doesn't belong here. I'd like to know why we need it and if needed we need to document this. 
This is not explicit why it's needed 🤔

atomrc commented on src/script/conversation/MessageRepository.ts:
```diff
@@ -892,8 +898,9 @@ export class MessageRepository {
-  private async sendReaction(conversation: Conversation, messageEntity: Message, reactionType: ReactionType) {
-    const reaction = MessageBuilder.buildReactionMessage({originalMessageId: messageEntity.id, type: reactionType});
+
+  private async sendReaction(conversation: Conversation, messageEntity: Message, reactions: ReactionType) {
```
The method name `reaction` (singular) but it takes a `reactions` (plural). That seems inconsistent

arjita-mitra replied:
reaction can be single or multiple comma separated. what do you suggest in this case?

atomrc replied:
Can we send multiple reactions at once? 
If that's the case, we should name it `sendReactions`. 

How is it possible to send multiple reactions at once? or does it overwrite each time we send?

atomrc commented on yalc.lock:
```diff
@@ -0,0 +1,10 @@
+{
+  "version": "v1",
+  "packages": {
+    "@wireapp/core": {
+      "signature": "ce501479df24a358c5bc9af495d7b8f4",
+      "file": true,
+      "replaced": "39.1.3"
+    }
+  }
+}
```
This should not be committed

arjita-mitra replied:
yes, i was planing to remove it after ui-kit version bump

atomrc commented on src/script/components/MessagesList/Message/MessageWrapper.tsx:
```diff
@@ -215,6 +215,7 @@ export const MessageWrapper: React.FC<MessageParams & {hasMarker: boolean; isMes
         onRetry={onRetry}
         isMessageFocused={isMessageFocused}
         isMsgElementsFocusable={isMsgElementsFocusable}
+        messageRepository={messageRepository}
```
```suggestion
        onClickReaction={() => messageRepository.sendReaction(...)}
```

arjita-mitra replied:
Not all message have reaction, only content message can have reaction. so it made more sense to me to have it inside contentmessage instead of message wrapper.

atomrc replied:
The problem is, this injects too much power in the `ContentMessage`. It should never have full access to the `MessageRepository` this is not going to scale. 
Only a handful of components should have access to it. In this case it's the `MessageWrapper`

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/MessageActions/MessageReactions/MessageReactionsList.test.tsx:
```diff
@@ -0,0 +1,77 @@
+
+import {MessageReactionsList} from './MessageReactionsList';
+
+describe('MessageReactionsList', () => {
```
🤩

arjita-mitra replied:
is there anything to fix here or you are just happy 🤔

atomrc replied:
Just happy :)

arjita-mitra committed:
chore: update yarn lock

atomrc commented on .env.localhost:
```diff
@@ -17,7 +17,7 @@ ENABLE_DEV_BACKEND_API="true"
 #FEATURE_APPLOCK_SCHEDULED_TIMEOUT="30"
 #FEATURE_ENABLE_MLS="true"
-#FEATURE_ENABLE_REACTION="false"
+#FEATURE_ENABLE_REACTION="true"
```
Is this consumed anywhere? 
Maybe this can be removed altogether now that we are developing the feature entirely in this branch (we don't need to keep the simple `like` around if we go this strategy)

arjita-mitra replied:
yes, it is being used inside ContentMessage.tsx  
const isReactionFeatureEnabled = Config.getConfig().FEATURE.ENABLE_REACTION;

we need the feature flag to disable the feature from being displayed on dev, after we merge.

thisisamir98 committed:
feat: Add ability to see messsage reactions (SQSERVICES-1876) (#14825)

* feat: Add ability to see messsage reactions (SQSERVICES-1876)

* getEmojiTitleFromEmojiUnicode

* feat: create util for transformReactionObj
replace thums up and heart emoji with image
image css
create list reactions component

* feat: Add emoji title utils (#14827)

* feat: convert emoji to unicode

* styles

* use first name

* use reduce

* use map

* use map for emoji dictionary

* remove extra ?

* remove duplicated code

* Update src/script/util/EmojiUtil.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* Update src/script/util/EmojiUtil.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* correct rendering

* yarn lock changes

* migrate styles

* dark mode handle

* change styles file name

---------

Co-authored-by: arjita-mitra <arjitamitra3@gmail.com>
Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

arjita-mitra committed:
fix: new and older client should be in sync when a user reacts on a message

arjita-mitra committed:
test: add unit testing for reaction add/remove from a message

arjita-mitra committed:
Merge branch 'feature/message-reaction' of https://github.com/wireapp/wire-webapp into feature/message-reaction

atomrc commented on src/script/components/MessagesList/Message/MessageWrapper.tsx:
```diff
@@ -189,6 +190,15 @@ export const MessageWrapper: React.FC<MessageParams & {hasMarker: boolean; isMes
+      return void messageRepository.toggleReaction(conversation, message, reaction, selfId.id);
+    },
+    [conversation, message, messageRepository, selfId.id],
+  );
```
I'd recommend not using `useCallback` and `useMemo` as default but rather when it's really needed. 
In this case it's a pure overhead

```suggestion
  const handleReactionClick = (reaction: ReactionType) => {
      if (!message.isContent()) {
        return;
      }
      return void messageRepository.toggleReaction(conversation, message, reaction, selfId.id);
    };
```

arjita-mitra committed:
Update src/script/components/MessagesList/Message/MessageWrapper.tsx

Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>
